File: cpp/src/0009.cpp
/*
[9. Palindrome Number (Easy)](https://leetcode.com/problems/palindrome-number/)

@topic: Math

Constraints:
  - `-2^31 <= x <= 2^31 - 1`
*/

class Solution {
   public:
    bool isPalindrome(int x) {
        if (x <= 9) return x >= 0;
        if (x % 10 == 0) return false;
        int p = 0;
        int q = x;
        while (q > p) {
            p = p * 10 + q % 10;
            q /= 10;
        }
        return p == q || p / 10 == q;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    Solution o;

    vector<tuple<int, bool>> CASES = {
        {121, true},
        {-121, false},
        {10, false},
    };

    for (auto& [x, excepted] : CASES) {
        assert(o.isPalindrome(x) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [x, _] : CASES) {
        for (int i = 0; i < 100000; ++i) {
            o.isPalindrome(x);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0009.py
class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x <= 9:
            return x >= 0
        if x % 10 == 0:
            return False
        p, q = 0, x
        while q > p:
            q, r = divmod(q, 10)
            p = p * 10 + r
        return p == q or p // 10 == q


if __name__ == '__main__':
    import unittest
    from timeit import timeit

    CASES = (
        (-1, False),
        (-10, False),
        (-11, False),
        (-111, False),
        (0, True),
        (1, True),
        (2, True),
        (3, True),
        (4, True),
        (5, True),
        (6, True),
        (7, True),
        (8, True),
        (9, True),
        (10, False),
        (121, True),
        (1001, True),
        (10201, True),
    )

    class Test(unittest.TestCase):
        def test(self):
            isPalindrome = Solution().isPalindrome
            for x, excepted in CASES:
                actual = isPalindrome(x)
                self.assertEqual(actual, excepted)

    unittest.main()

    isPalindrome = Solution().isPalindrome
    for x, _ in CASES:
        print(timeit(lambda: isPalindrome(x), number=10000))
